In [2]:
import numpy as np
import pandas as pd

In [3]:
match=pd.read_csv(r"C:\Users\Soumadeep\Desktop\WN PROBABILITY PREDICTER\IPL_Matches_2008_2022.csv")
delivery=pd.read_csv(r"C:\Users\Soumadeep\Desktop\WN PROBABILITY PREDICTER\IPL_Ball_by_Ball_2008_2022.csv")


In [4]:
match.head()

ID       City        Date Season  MatchNumber  \
0  1312200  Ahmedabad  2022-05-29   2022        Final   
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   
2  1312198    Kolkata  2022-05-25   2022   Eliminator   
3  1312197    Kolkata  2022-05-24   2022  Qualifier 1   
4  1304116     Mumbai  2022-05-22   2022           70   

                         Team1                 Team2  \
0             Rajasthan Royals        Gujarat Titans   
1  Royal Challengers Bangalore      Rajasthan Royals   
2  Royal Challengers Bangalore  Lucknow Super Giants   
3             Rajasthan Royals        Gujarat Titans   
4          Sunrisers Hyderabad          Punjab Kings   

                              Venue            TossWinner TossDecision  \
0  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals          bat   
1  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals        field   
2             Eden Gardens, Kolkata  Lucknow Super Giants        field   
3             Eden Gardens, Kolkata        Gujarat Titans        field   
4          Wankhede Stadium, Mumbai   Sunrisers Hyderabad          bat   

  SuperOver                  WinningTeam    WonBy  Margin method  \
0         N               Gujarat Titans  Wickets     7.0    NaN   
1         N             Rajasthan Royals  Wickets     7.0    NaN   
2         N  Royal Challengers Bangalore     Runs    14.0    NaN   
3         N               Gujarat Titans  Wickets     7.0    NaN   
4         N                 Punjab Kings  Wickets     5.0    NaN   

  Player_of_Match                                       Team1Players  \
0       HH Pandya  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1      JC Buttler  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
2      RM Patidar  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
3       DA Miller  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
4   Harpreet Brar  ['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...   

                                        Team2Players        Umpire1  \
0  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...    CB Gaffaney   
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...    CB Gaffaney   
2  ['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...  J Madanagopal   
3  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...   BNJ Oxenford   
4  ['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...   AK Chaudhary   

         Umpire2  
0    Nitin Menon  
1    Nitin Menon  
2       MA Gough  
3      VK Sharma  
4  NA Patwardhan

In [5]:
#CHECKING THE DIMENTION OF THE MATCH DATASET
match.shape

(950, 20)

In [6]:
delivery.head()

ID  innings  overs  ballnumber       batter          bowler  \
0  1312200        1      0           1  YBK Jaiswal  Mohammed Shami   
1  1312200        1      0           2  YBK Jaiswal  Mohammed Shami   
2  1312200        1      0           3   JC Buttler  Mohammed Shami   
3  1312200        1      0           4  YBK Jaiswal  Mohammed Shami   
4  1312200        1      0           5  YBK Jaiswal  Mohammed Shami   

   non-striker extra_type  batsman_run  extras_run  total_run  non_boundary  \
0   JC Buttler        NaN            0           0          0             0   
1   JC Buttler    legbyes            0           1          1             0   
2  YBK Jaiswal        NaN            1           0          1             0   
3   JC Buttler        NaN            0           0          0             0   
4   JC Buttler        NaN            0           0          0             0   

   isWicketDelivery player_out kind fielders_involved       BattingTeam  
0                 0        NaN  NaN               NaN  Rajasthan Royals  
1                 0        NaN  NaN               NaN  Rajasthan Royals  
2                 0        NaN  NaN               NaN  Rajasthan Royals  
3                 0        NaN  NaN               NaN  Rajasthan Royals  
4                 0        NaN  NaN               NaN  Rajasthan Royals

In [7]:
# CHECKING THE DIMENTION OF DELIVERY DATASET
delivery.shape

(225954, 17)

# NOW THIS IS A CLASSIFICATION PROBLEM BUT WE HAVE THE CHOOSE THAT TYPE OF MODEL WHICH CAN GIVE RESULT IN PROBABILITY .....
# AGAIN WE DONT NEED THESE VAST DATA ....
# WE HAVE TO EXTRACT OUR NECESSARY INFORMATION FROM THOSE TWO DATASETS ON WHICH OUR MODEL ACTUALLY RUNS

In [8]:
# WE WANT TO DETERMINE THE TOTAL RUNS SCORED IN BOTH INNINGS
total_score_df=delivery.groupby(['ID','innings']).sum()['total_run'].reset_index()
total_score_df.head()

C:\Users\Soumadeep\AppData\Local\Temp\ipykernel_4860\366049979.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  total_score_df=delivery.groupby(['ID','innings']).sum()['total_run'].reset_index()


ID  innings  total_run
0  335982        1        222
1  335982        2         82
2  335983        1        240
3  335983        2        207
4  335984        1        129

In [9]:
#now we dont need the 2nd innings total score as we only want the first innings score which the 2nd team are chasing
total_score_df=total_score_df[total_score_df['innings']==1]
total_score_df.head()

ID  innings  total_run
0  335982        1        222
2  335983        1        240
4  335984        1        129
6  335985        1        165
8  335986        1        110

In [10]:
#now we merge this to our match dataset as in that dataset all the match details are there but the 1st innings total are not there
match_df=match.merge(total_score_df[['ID','total_run']])
# here we don't use any left join or right join condition as two data frame has a common coloumn name 'ID'

In [11]:
match_df.head()

ID       City        Date Season  MatchNumber  \
0  1312200  Ahmedabad  2022-05-29   2022        Final   
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   
2  1312198    Kolkata  2022-05-25   2022   Eliminator   
3  1312197    Kolkata  2022-05-24   2022  Qualifier 1   
4  1304116     Mumbai  2022-05-22   2022           70   

                         Team1                 Team2  \
0             Rajasthan Royals        Gujarat Titans   
1  Royal Challengers Bangalore      Rajasthan Royals   
2  Royal Challengers Bangalore  Lucknow Super Giants   
3             Rajasthan Royals        Gujarat Titans   
4          Sunrisers Hyderabad          Punjab Kings   

                              Venue            TossWinner TossDecision  ...  \
0  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals          bat  ...   
1  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals        field  ...   
2             Eden Gardens, Kolkata  Lucknow Super Giants        field  ...   
3             Eden Gardens, Kolkata        Gujarat Titans        field  ...   
4          Wankhede Stadium, Mumbai   Sunrisers Hyderabad          bat  ...   

                   WinningTeam    WonBy Margin  method Player_of_Match  \
0               Gujarat Titans  Wickets    7.0     NaN       HH Pandya   
1             Rajasthan Royals  Wickets    7.0     NaN      JC Buttler   
2  Royal Challengers Bangalore     Runs   14.0     NaN      RM Patidar   
3               Gujarat Titans  Wickets    7.0     NaN       DA Miller   
4                 Punjab Kings  Wickets    5.0     NaN   Harpreet Brar   

                                        Team1Players  \
0  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
2  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
3  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
4  ['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...   

                                        Team2Players        Umpire1  \
0  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...    CB Gaffaney   
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...    CB Gaffaney   
2  ['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...  J Madanagopal   
3  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...   BNJ Oxenford   
4  ['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...   AK Chaudhary   

         Umpire2 total_run  
0    Nitin Menon       130  
1    Nitin Menon       157  
2       MA Gough       207  
3      VK Sharma       188  
4  NA Patwardhan       157  

[5 rows x 21 columns]

In [12]:
# THERE ARE SUCH TEAMS WHICH DONT PLAY ANYMORE IN THE IPL 
# HERE WE NEED ONLY THOSE TEAMS WHICH CONTINUOUSLY PLAYING & ARE ACTIVE MODE RIGHT NOW
match_df['Team1'].unique()

array(['Rajasthan Royals', 'Royal Challengers Bangalore',
       'Sunrisers Hyderabad', 'Delhi Capitals', 'Chennai Super Kings',
       'Gujarat Titans', 'Lucknow Super Giants', 'Kolkata Knight Riders',
       'Punjab Kings', 'Mumbai Indians', 'Kings XI Punjab',
       'Delhi Daredevils', 'Rising Pune Supergiant', 'Gujarat Lions',
       'Rising Pune Supergiants', 'Pune Warriors', 'Deccan Chargers',
       'Kochi Tuskers Kerala'], dtype=object)

In [13]:
#list of teams which are currently active
teams=['Sunrisers Hyderabad','Mumbai Indians','Royal Challengers Bangalore','Kolkata Knight Riders','Punjab Kings','Chennai Super Kings','Rajasthan Royals','Delhi Capitals']

In [14]:
# now we change the team names which change their names after some year of playing IPL
match_df['Team1']=match_df['Team1'].str.replace('Delhi Daredevils','Delhi Capitals')
match_df['Team2']=match_df['Team2'].str.replace('Delhi Daredevils','Delhi Capitals')

match_df['Team1']=match_df['Team1'].str.replace('Deccan Chargers','Sunrisers Hyderabad')
match_df['Team2']=match_df['Team2'].str.replace('Deccan Chargers','Sunrisers Hyderabad')

match_df['Team1']=match_df['Team1'].str.replace('Kings XI Punjab','Punjab Kings')
match_df['Team2']=match_df['Team2'].str.replace('Kings XI Punjab','Punjab Kings')



In [15]:
# we want to take those rows where the teams are in our teams list
match_df=match_df[match_df['Team1'].isin(teams)]
match_df=match_df[match_df['Team2'].isin(teams)]

In [16]:
match_df.shape

(806, 21)

# WE CAN SEE THAT, AT FIRST WE HAVE ABOUT 950 MATCHES PRESENT IN THE MATCH DATASET.
# BUT NOW AS WE CLEAN THE DATASET WE GET ONLY 806 MATCHES

In [17]:
# now we extract our necessary coloums from this match dataset & merge to the delivery dataset as the main features are present in the delivery dataset
match_df.head()

ID       City        Date Season  MatchNumber  \
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   
4  1304116     Mumbai  2022-05-22   2022           70   
5  1304115     Mumbai  2022-05-21   2022           69   
6  1304114     Mumbai  2022-05-20   2022           68   
9  1304111     Mumbai  2022-05-17   2022           65   

                         Team1             Team2  \
1  Royal Challengers Bangalore  Rajasthan Royals   
4          Sunrisers Hyderabad      Punjab Kings   
5               Delhi Capitals    Mumbai Indians   
6          Chennai Super Kings  Rajasthan Royals   
9          Sunrisers Hyderabad    Mumbai Indians   

                              Venue           TossWinner TossDecision  ...  \
1  Narendra Modi Stadium, Ahmedabad     Rajasthan Royals        field  ...   
4          Wankhede Stadium, Mumbai  Sunrisers Hyderabad          bat  ...   
5          Wankhede Stadium, Mumbai       Mumbai Indians        field  ...   
6         Brabourne Stadium, Mumbai  Chennai Super Kings          bat  ...   
9          Wankhede Stadium, Mumbai       Mumbai Indians        field  ...   

           WinningTeam    WonBy Margin  method Player_of_Match  \
1     Rajasthan Royals  Wickets    7.0     NaN      JC Buttler   
4         Punjab Kings  Wickets    5.0     NaN   Harpreet Brar   
5       Mumbai Indians  Wickets    5.0     NaN       JJ Bumrah   
6     Rajasthan Royals  Wickets    5.0     NaN        R Ashwin   
9  Sunrisers Hyderabad     Runs    3.0     NaN     RA Tripathi   

                                        Team1Players  \
1  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
4  ['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...   
5  ['PP Shaw', 'DA Warner', 'MR Marsh', 'RR Pant'...   
6  ['RD Gaikwad', 'DP Conway', 'MM Ali', 'N Jagad...   
9  ['Abhishek Sharma', 'PK Garg', 'RA Tripathi', ...   

                                        Team2Players       Umpire1  \
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   CB Gaffaney   
4  ['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...  AK Chaudhary   
5  ['Ishan Kishan', 'RG Sharma', 'D Brevis', 'Til...   Nitin Menon   
6  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   CB Gaffaney   
9  ['RG Sharma', 'Ishan Kishan', 'DR Sams', 'Tila...   CB Gaffaney   

         Umpire2 total_run  
1    Nitin Menon       157  
4  NA Patwardhan       157  
5   Tapan Sharma       159  
6  NA Patwardhan       150  
9       N Pandit       193  

[5 rows x 21 columns]

# NOW WE NEED THE FOLLOWING COLOUMNS ONLY:
# ID , CITY , WINNING TEAM , TOTAL RUN , TEAM 1

In [18]:
match_df=match_df[['ID','City','WinningTeam','total_run','Team1']]
# WE MERGE THESE COLOUMNS WITH DELIVERY DATASET
delivery_df=match_df.merge(delivery,on='ID')
delivery_df.head()

ID       City       WinningTeam  total_run_x  \
0  1312199  Ahmedabad  Rajasthan Royals          157   
1  1312199  Ahmedabad  Rajasthan Royals          157   
2  1312199  Ahmedabad  Rajasthan Royals          157   
3  1312199  Ahmedabad  Rajasthan Royals          157   
4  1312199  Ahmedabad  Rajasthan Royals          157   

                         Team1  innings  overs  ballnumber        batter  \
0  Royal Challengers Bangalore        1      0           1       V Kohli   
1  Royal Challengers Bangalore        1      0           2       V Kohli   
2  Royal Challengers Bangalore        1      0           3  F du Plessis   
3  Royal Challengers Bangalore        1      0           4       V Kohli   
4  Royal Challengers Bangalore        1      0           5       V Kohli   

     bowler  ... extra_type batsman_run  extras_run  total_run_y  \
0  TA Boult  ...        NaN           0           0            0   
1  TA Boult  ...        NaN           1           0            1   
2  TA Boult  ...        NaN           1           0            1   
3  TA Boult  ...        NaN           0           0            0   
4  TA Boult  ...        NaN           0           0            0   

   non_boundary  isWicketDelivery  player_out kind fielders_involved  \
0             0                 0         NaN  NaN               NaN   
1             0                 0         NaN  NaN               NaN   
2             0                 0         NaN  NaN               NaN   
3             0                 0         NaN  NaN               NaN   
4             0                 0         NaN  NaN               NaN   

                   BattingTeam  
0  Royal Challengers Bangalore  
1  Royal Challengers Bangalore  
2  Royal Challengers Bangalore  
3  Royal Challengers Bangalore  
4  Royal Challengers Bangalore  

[5 rows x 21 columns]

In [19]:
# now we are only concerned with ball by ball data of 2nd innings of all matches
# thats why we eclude all the rows where the innings is equals to 1
delivery_df=delivery_df[delivery_df['innings']==2]
delivery.shape

(225954, 17)

In [20]:
# now we want the cuurent score per match wise & which updates in every ball
delivery_df['current_score']=delivery_df.groupby('ID').cumsum()['total_run_y']
delivery_df.head()

C:\Users\Soumadeep\AppData\Local\Temp\ipykernel_4860\3483366994.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.cumsum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  delivery_df['current_score']=delivery_df.groupby('ID').cumsum()['total_run_y']


ID       City       WinningTeam  total_run_x  \
127  1312199  Ahmedabad  Rajasthan Royals          157   
128  1312199  Ahmedabad  Rajasthan Royals          157   
129  1312199  Ahmedabad  Rajasthan Royals          157   
130  1312199  Ahmedabad  Rajasthan Royals          157   
131  1312199  Ahmedabad  Rajasthan Royals          157   

                           Team1  innings  overs  ballnumber       batter  \
127  Royal Challengers Bangalore        2      0           1  YBK Jaiswal   
128  Royal Challengers Bangalore        2      0           2  YBK Jaiswal   
129  Royal Challengers Bangalore        2      0           3  YBK Jaiswal   
130  Royal Challengers Bangalore        2      0           4  YBK Jaiswal   
131  Royal Challengers Bangalore        2      0           5  YBK Jaiswal   

             bowler  ... batsman_run extras_run  total_run_y  non_boundary  \
127  Mohammed Siraj  ...           0          0            0             0   
128  Mohammed Siraj  ...           0          0            0             0   
129  Mohammed Siraj  ...           6          0            6             0   
130  Mohammed Siraj  ...           4          0            4             0   
131  Mohammed Siraj  ...           0          0            0             0   

     isWicketDelivery  player_out  kind fielders_involved       BattingTeam  \
127                 0         NaN   NaN               NaN  Rajasthan Royals   
128                 0         NaN   NaN               NaN  Rajasthan Royals   
129                 0         NaN   NaN               NaN  Rajasthan Royals   
130                 0         NaN   NaN               NaN  Rajasthan Royals   
131                 0         NaN   NaN               NaN  Rajasthan Royals   

    current_score  
127             0  
128             0  
129             6  
130            10  
131            10  

[5 rows x 22 columns]

In [21]:
# here we want to update the remaining runs after every ball
delivery_df['target']=delivery_df['total_run_x']+1
delivery_df['runs_left']=delivery_df['target']-delivery_df['current_score']
delivery_df.head()

ID       City       WinningTeam  total_run_x  \
127  1312199  Ahmedabad  Rajasthan Royals          157   
128  1312199  Ahmedabad  Rajasthan Royals          157   
129  1312199  Ahmedabad  Rajasthan Royals          157   
130  1312199  Ahmedabad  Rajasthan Royals          157   
131  1312199  Ahmedabad  Rajasthan Royals          157   

                           Team1  innings  overs  ballnumber       batter  \
127  Royal Challengers Bangalore        2      0           1  YBK Jaiswal   
128  Royal Challengers Bangalore        2      0           2  YBK Jaiswal   
129  Royal Challengers Bangalore        2      0           3  YBK Jaiswal   
130  Royal Challengers Bangalore        2      0           4  YBK Jaiswal   
131  Royal Challengers Bangalore        2      0           5  YBK Jaiswal   

             bowler  ... total_run_y non_boundary  isWicketDelivery  \
127  Mohammed Siraj  ...           0            0                 0   
128  Mohammed Siraj  ...           0            0                 0   
129  Mohammed Siraj  ...           6            0                 0   
130  Mohammed Siraj  ...           4            0                 0   
131  Mohammed Siraj  ...           0            0                 0   

     player_out  kind  fielders_involved       BattingTeam current_score  \
127         NaN   NaN                NaN  Rajasthan Royals             0   
128         NaN   NaN                NaN  Rajasthan Royals             0   
129         NaN   NaN                NaN  Rajasthan Royals             6   
130         NaN   NaN                NaN  Rajasthan Royals            10   
131         NaN   NaN                NaN  Rajasthan Royals            10   

    target runs_left  
127    158       158  
128    158       158  
129    158       152  
130    158       148  
131    158       148  

[5 rows x 24 columns]

In [22]:
#now we want the number of remaining balls after every ball was pitched
delivery_df['remaining_balls']=120-(delivery_df['overs']*6+delivery_df['ballnumber'])
delivery_df.head()

ID       City       WinningTeam  total_run_x  \
127  1312199  Ahmedabad  Rajasthan Royals          157   
128  1312199  Ahmedabad  Rajasthan Royals          157   
129  1312199  Ahmedabad  Rajasthan Royals          157   
130  1312199  Ahmedabad  Rajasthan Royals          157   
131  1312199  Ahmedabad  Rajasthan Royals          157   

                           Team1  innings  overs  ballnumber       batter  \
127  Royal Challengers Bangalore        2      0           1  YBK Jaiswal   
128  Royal Challengers Bangalore        2      0           2  YBK Jaiswal   
129  Royal Challengers Bangalore        2      0           3  YBK Jaiswal   
130  Royal Challengers Bangalore        2      0           4  YBK Jaiswal   
131  Royal Challengers Bangalore        2      0           5  YBK Jaiswal   

             bowler  ... non_boundary isWicketDelivery  player_out  kind  \
127  Mohammed Siraj  ...            0                0         NaN   NaN   
128  Mohammed Siraj  ...            0                0         NaN   NaN   
129  Mohammed Siraj  ...            0                0         NaN   NaN   
130  Mohammed Siraj  ...            0                0         NaN   NaN   
131  Mohammed Siraj  ...            0                0         NaN   NaN   

     fielders_involved       BattingTeam  current_score target runs_left  \
127                NaN  Rajasthan Royals              0    158       158   
128                NaN  Rajasthan Royals              0    158       158   
129                NaN  Rajasthan Royals              6    158       152   
130                NaN  Rajasthan Royals             10    158       148   
131                NaN  Rajasthan Royals             10    158       148   

    remaining_balls  
127             119  
128             118  
129             117  
130             116  
131             115  

[5 rows x 25 columns]

In [23]:
# now we want the number of remaining wickets 
wickets=delivery_df.groupby('ID').cumsum()['isWicketDelivery'].values
delivery_df['remaining_wickets']=10-wickets
delivery_df.tail()

C:\Users\Soumadeep\AppData\Local\Temp\ipykernel_4860\144419026.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.cumsum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  wickets=delivery_df.groupby('ID').cumsum()['isWicketDelivery'].values


ID       City            WinningTeam  total_run_x  \
191826  335982  Bangalore  Kolkata Knight Riders          222   
191827  335982  Bangalore  Kolkata Knight Riders          222   
191828  335982  Bangalore  Kolkata Knight Riders          222   
191829  335982  Bangalore  Kolkata Knight Riders          222   
191830  335982  Bangalore  Kolkata Knight Riders          222   

                              Team1  innings  overs  ballnumber    batter  \
191826  Royal Challengers Bangalore        2     14           5   P Kumar   
191827  Royal Challengers Bangalore        2     14           6  SB Joshi   
191828  Royal Challengers Bangalore        2     14           7   P Kumar   
191829  Royal Challengers Bangalore        2     15           1  SB Joshi   
191830  Royal Challengers Bangalore        2     15           2  SB Joshi   

           bowler  ... isWicketDelivery player_out    kind  fielders_involved  \
191826   I Sharma  ...                0        NaN     NaN                NaN   
191827   I Sharma  ...                0        NaN     NaN                NaN   
191828   I Sharma  ...                0        NaN     NaN                NaN   
191829  LR Shukla  ...                0        NaN     NaN                NaN   
191830  LR Shukla  ...                1   SB Joshi  caught        BB McCullum   

                        BattingTeam  current_score  target runs_left  \
191826  Royal Challengers Bangalore             80     223       143   
191827  Royal Challengers Bangalore             81     223       142   
191828  Royal Challengers Bangalore             81     223       142   
191829  Royal Challengers Bangalore             82     223       141   
191830  Royal Challengers Bangalore             82     223       141   

       remaining_balls remaining_wickets  
191826              31                 1  
191827              30                 1  
191828              29                 1  
191829              29                 1  
191830              28                 0  

[5 rows x 26 columns]

In [24]:
# NOW WE WANT TO CALCULATE CURRENT RUN RATE (CRR)
# CRR = RUNS/OVER
# crr = current score/[(120-remaining balls)/6]=current score * 6 / (120 - remaining balls)
delivery_df['crr']=delivery_df['current_score']*6/(120-delivery_df['remaining_balls'])
delivery_df.tail()

ID       City            WinningTeam  total_run_x  \
191826  335982  Bangalore  Kolkata Knight Riders          222   
191827  335982  Bangalore  Kolkata Knight Riders          222   
191828  335982  Bangalore  Kolkata Knight Riders          222   
191829  335982  Bangalore  Kolkata Knight Riders          222   
191830  335982  Bangalore  Kolkata Knight Riders          222   

                              Team1  innings  overs  ballnumber    batter  \
191826  Royal Challengers Bangalore        2     14           5   P Kumar   
191827  Royal Challengers Bangalore        2     14           6  SB Joshi   
191828  Royal Challengers Bangalore        2     14           7   P Kumar   
191829  Royal Challengers Bangalore        2     15           1  SB Joshi   
191830  Royal Challengers Bangalore        2     15           2  SB Joshi   

           bowler  ... player_out    kind  fielders_involved  \
191826   I Sharma  ...        NaN     NaN                NaN   
191827   I Sharma  ...        NaN     NaN                NaN   
191828   I Sharma  ...        NaN     NaN                NaN   
191829  LR Shukla  ...        NaN     NaN                NaN   
191830  LR Shukla  ...   SB Joshi  caught        BB McCullum   

                        BattingTeam  current_score  target  runs_left  \
191826  Royal Challengers Bangalore             80     223        143   
191827  Royal Challengers Bangalore             81     223        142   
191828  Royal Challengers Bangalore             81     223        142   
191829  Royal Challengers Bangalore             82     223        141   
191830  Royal Challengers Bangalore             82     223        141   

       remaining_balls remaining_wickets       crr  
191826              31                 1  5.393258  
191827              30                 1  5.400000  
191828              29                 1  5.340659  
191829              29                 1  5.406593  
191830              28                 0  5.347826  

[5 rows x 27 columns]

In [25]:
# now we want to calculate required run rate
# required run rate = runs left / (remaining balls/6) = ( runs left * 6 ) / remaining balls
delivery_df['rrr']=(delivery_df['runs_left']*6)/delivery_df['remaining_balls']
delivery_df.head()

ID       City       WinningTeam  total_run_x  \
127  1312199  Ahmedabad  Rajasthan Royals          157   
128  1312199  Ahmedabad  Rajasthan Royals          157   
129  1312199  Ahmedabad  Rajasthan Royals          157   
130  1312199  Ahmedabad  Rajasthan Royals          157   
131  1312199  Ahmedabad  Rajasthan Royals          157   

                           Team1  innings  overs  ballnumber       batter  \
127  Royal Challengers Bangalore        2      0           1  YBK Jaiswal   
128  Royal Challengers Bangalore        2      0           2  YBK Jaiswal   
129  Royal Challengers Bangalore        2      0           3  YBK Jaiswal   
130  Royal Challengers Bangalore        2      0           4  YBK Jaiswal   
131  Royal Challengers Bangalore        2      0           5  YBK Jaiswal   

             bowler  ... kind fielders_involved       BattingTeam  \
127  Mohammed Siraj  ...  NaN               NaN  Rajasthan Royals   
128  Mohammed Siraj  ...  NaN               NaN  Rajasthan Royals   
129  Mohammed Siraj  ...  NaN               NaN  Rajasthan Royals   
130  Mohammed Siraj  ...  NaN               NaN  Rajasthan Royals   
131  Mohammed Siraj  ...  NaN               NaN  Rajasthan Royals   

     current_score  target  runs_left  remaining_balls remaining_wickets  \
127              0     158        158              119                10   
128              0     158        158              118                10   
129              6     158        152              117                10   
130             10     158        148              116                10   
131             10     158        148              115                10   

      crr       rrr  
127   0.0  7.966387  
128   0.0  8.033898  
129  12.0  7.794872  
130  15.0  7.655172  
131  12.0  7.721739  

[5 rows x 28 columns]

# now we want to create a coloumn named result which shows that if the 2nd batting team is winner or not
# if the chasing team wins then it takes the value 1 
# otherwise 0
# so we are doing this with the help of creating a function

In [26]:
def result(row):
    return 1 if row['BattingTeam']==row['WinningTeam'] else 0

In [27]:
delivery_df['result']= delivery_df.apply(result,axis=1)
delivery_df.head()

ID       City       WinningTeam  total_run_x  \
127  1312199  Ahmedabad  Rajasthan Royals          157   
128  1312199  Ahmedabad  Rajasthan Royals          157   
129  1312199  Ahmedabad  Rajasthan Royals          157   
130  1312199  Ahmedabad  Rajasthan Royals          157   
131  1312199  Ahmedabad  Rajasthan Royals          157   

                           Team1  innings  overs  ballnumber       batter  \
127  Royal Challengers Bangalore        2      0           1  YBK Jaiswal   
128  Royal Challengers Bangalore        2      0           2  YBK Jaiswal   
129  Royal Challengers Bangalore        2      0           3  YBK Jaiswal   
130  Royal Challengers Bangalore        2      0           4  YBK Jaiswal   
131  Royal Challengers Bangalore        2      0           5  YBK Jaiswal   

             bowler  ... fielders_involved       BattingTeam  current_score  \
127  Mohammed Siraj  ...               NaN  Rajasthan Royals              0   
128  Mohammed Siraj  ...               NaN  Rajasthan Royals              0   
129  Mohammed Siraj  ...               NaN  Rajasthan Royals              6   
130  Mohammed Siraj  ...               NaN  Rajasthan Royals             10   
131  Mohammed Siraj  ...               NaN  Rajasthan Royals             10   

     target  runs_left  remaining_balls  remaining_wickets   crr       rrr  \
127     158        158              119                 10   0.0  7.966387   
128     158        158              118                 10   0.0  8.033898   
129     158        152              117                 10  12.0  7.794872   
130     158        148              116                 10  15.0  7.655172   
131     158        148              115                 10  12.0  7.721739   

    result  
127      1  
128      1  
129      1  
130      1  
131      1  

[5 rows x 29 columns]

In [28]:
final_df=delivery_df[['City','target','Team1','BattingTeam','runs_left','remaining_balls','remaining_wickets','crr','rrr','result']]

                     
            

In [29]:
final_df.head()

City  target                        Team1       BattingTeam  \
127  Ahmedabad     158  Royal Challengers Bangalore  Rajasthan Royals   
128  Ahmedabad     158  Royal Challengers Bangalore  Rajasthan Royals   
129  Ahmedabad     158  Royal Challengers Bangalore  Rajasthan Royals   
130  Ahmedabad     158  Royal Challengers Bangalore  Rajasthan Royals   
131  Ahmedabad     158  Royal Challengers Bangalore  Rajasthan Royals   

     runs_left  remaining_balls  remaining_wickets   crr       rrr  result  
127        158              119                 10   0.0  7.966387       1  
128        158              118                 10   0.0  8.033898       1  
129        152              117                 10  12.0  7.794872       1  
130        148              116                 10  15.0  7.655172       1  
131        148              115                 10  12.0  7.721739       1

# NOW WE HAVE ALL THE BALLS PLAYED IN ONE MATCH TOGETHER
# IT CAN CREATE A BIAS 
# THAT'S WHY WE SHUFFLE OUR FINAL DATADET

In [30]:
# we shuffule the final dataset
final_df=final_df.sample(final_df.shape[0])

In [31]:
final_df.sample()
#it shows any row of the final dataset randomly

City  target                        Team1      BattingTeam  \
165619  Centurion     171  Royal Challengers Bangalore  Deccan Chargers   

        runs_left  remaining_balls  remaining_wickets        crr       rrr  \
165619        105               81                 10  10.153846  7.777778   

        result  
165619       0

In [32]:
final_df.isnull().sum()

City                 6012
target                  0
Team1                   0
BattingTeam             0
runs_left               0
remaining_balls         0
remaining_wickets       0
crr                     0
rrr                    10
result                  0
dtype: int64

# we see there is some null values present in the final dataset
# So we drop all of them

In [33]:
final_df.dropna(inplace=True)
# We drop all the null Values

In [34]:
final_df.describe()

target     runs_left  remaining_balls  remaining_wickets  \
count  86464.000000  86464.000000     86464.000000       86464.000000   
mean     165.630019     92.704189        62.898166           7.550287   
std       29.099192     49.747764        33.392094           2.148235   
min       63.000000    -15.000000        -2.000000           0.000000   
25%      148.000000     53.000000        35.000000           6.000000   
50%      166.000000     92.000000        64.000000           8.000000   
75%      185.000000    130.000000        92.000000           9.000000   
max      247.000000    247.000000       119.000000          10.000000   

                crr           rrr        result  
count  86464.000000  8.646400e+04  86464.000000  
mean       7.443720           NaN      0.530082  
std        2.271445           NaN      0.499097  
min        0.000000          -inf      0.000000  
25%        6.276923  7.177570e+00      0.000000  
50%        7.500000  8.865672e+00      1.000000  
75%        8.666667  1.095000e+01      1.000000  
max       36.000000           inf      1.000000

# We easily see that in the Required Run Rate Column there is still some Nan values , some Infinity values & some large values are present 
# We have to handle Them before Proceeding

# We can do one thing
# we drop those rows where remaining balls are equal to 0
# Because It creates Problem of taking Infinite Values

In [35]:
# taking those rows only where remaining balls are not equal to 0
final_df=final_df[final_df['remaining_balls']!=0]

# NOW WE CAN CONCENTRATE ON MODEL BUILDING
# DATA CLEANING PART IS DONE


In [36]:
# WE SPLIT TRAIN & TEST SET
x=final_df.drop(['result'],axis=1)
y=final_df['result']
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=18)

In [37]:
x_train

City  target                        Team1  \
64777   Chandigarh     194                 Punjab Kings   
92401    Bangalore     182  Royal Challengers Bangalore   
182530   Hyderabad     179          Sunrisers Hyderabad   
98642      Kolkata     161        Kolkata Knight Riders   
137730   Bangalore     166  Royal Challengers Bangalore   
...            ...     ...                          ...   
54749      Kolkata     170        Kolkata Knight Riders   
134362      Mumbai     164               Mumbai Indians   
113404      Mumbai     167               Mumbai Indians   
142597      Jaipur     147             Rajasthan Royals   
177079   Cape Town     185  Royal Challengers Bangalore   

                        BattingTeam  runs_left  remaining_balls  \
64777           Sunrisers Hyderabad        154               84   
92401   Royal Challengers Bangalore         58               17   
182530              Deccan Chargers        131               72   
98642         Kolkata Knight Riders         90               66   
137730  Royal Challengers Bangalore         52                8   
...                             ...        ...              ...   
54749              Rajasthan Royals        121               87   
134362              Kings XI Punjab         55               34   
113404             Rajasthan Royals         38               19   
142597  Royal Challengers Bangalore         31               32   
177079  Royal Challengers Bangalore        147               79   

        remaining_wickets       crr        rrr  
64777                   7  6.666667  11.000000  
92401                   3  7.223301  20.470588  
182530                  7  6.000000  10.916667  
98642                   8  7.888889   8.181818  
137730                  2  6.107143  39.000000  
...                   ...       ...        ...  
54749                   9  8.909091   8.344828  
134362                  8  7.604651   9.705882  
113404                  4  7.663366  12.000000  
142597                  9  7.909091   5.812500  
177079                  8  5.560976  11.164557  

[64642 rows x 9 columns]

# now we can see that in our final dataset there are 3 coloumns which are not numerical in nature
# which can cause a problem to us
# that's why we have to apply ONE-HOT encoding method to convert those columns to numerical nature

In [38]:
#one-hot encoding
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
trf=ColumnTransformer([('trf',OneHotEncoder(sparse=False,drop='first'),['BattingTeam','Team1','City'])],remainder='passthrough')

In [39]:
# NOW WE APPLY OUR LOGISTIC REGRESSION MODEL
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
# WE USE PIPELINE BY THIS


In [40]:
# We created the Pipeline
pipe=Pipeline(steps=[('step1',trf),('step2',LogisticRegression(solver='liblinear'))])

In [41]:
# PREDICT TIME!!
pipe.fit(x_train,y_train)

C:\Users\Soumadeep\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Pipeline(steps=[('step1',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('trf',
                                                  OneHotEncoder(drop='first',
                                                                sparse=False),
                                                  ['BattingTeam', 'Team1',
                                                   'City'])])),
                ('step2', LogisticRegression(solver='liblinear'))])

In [42]:
y_pred=pipe.predict(x_test)

In [43]:
# CHECKING ACCURACY OF THE MODEL
from sklearn.metrics import accuracy_score
a=accuracy_score(y_test,y_pred)

In [44]:
print('The Acuuracy of Our Model:','%.2f'%(a))

The Acuuracy of Our Model: 0.81


# now we can do Random forest also 
# but it will give our desired probability in any situation too biased to one team specifically
# But logistic Regression will give some smooth values rather than Random Forest

# Now our model is ready 
# we just put our model into a website


In [45]:
# At first we want the name of all teams
teams

['Sunrisers Hyderabad',
 'Mumbai Indians',
 'Royal Challengers Bangalore',
 'Kolkata Knight Riders',
 'Punjab Kings',
 'Chennai Super Kings',
 'Rajasthan Royals',
 'Delhi Capitals']

In [46]:
#Then we need All the unique Cities 
delivery_df['City'].unique()

array(['Ahmedabad', 'Mumbai', 'Navi Mumbai', 'Pune', 'Dubai', 'Sharjah',
       'Abu Dhabi', 'Delhi', 'Chennai', nan, 'Hyderabad', 'Visakhapatnam',
       'Chandigarh', 'Bengaluru', 'Kolkata', 'Jaipur', 'Indore',
       'Bangalore', 'Raipur', 'Ranchi', 'Cuttack', 'Dharamsala', 'Nagpur',
       'Johannesburg', 'Centurion', 'Durban', 'Bloemfontein',
       'Port Elizabeth', 'Kimberley', 'East London', 'Cape Town'],
      dtype=object)

In [47]:
import pickle
pickle.dump(pipe,open('pipe.pkl','wb'))

In [52]:
input1_df=pd.DataFrame({'BattingTeam':['Chennai Super Kings'],'Team1':['Punjab Kings'],
                           'City':['Chandigarh'],'runs_left':[100],
                           'remaining_balls':[60],'remaining_wickets':[5],
                           'target':[200],
                           'crr':[10],'rrr':[10]})

In [53]:
input1_df


BattingTeam         Team1        City  runs_left  remaining_balls  \
0  Chennai Super Kings  Punjab Kings  Chandigarh        100               60   

   remaining_wickets  target  crr  rrr  
0                  5     200   10   10

In [55]:
pipe.predict_proba(input1_df)

array([[0.77370736, 0.22629264]])